crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -1122,6 +1150,187 @@ mod tests {
+            )
+            .await;
+
+        assert_matches!(
+            request_result,
+            Err(OlmError::SessionRecipientCollectionError(
+                SessionRecipientCollectionError::CrossSigningNotSetup
+            ))
+        );
+
+        // We now get our public cross-signing keys, but we don't trust them
+        // yet.  In this case, sharing the keys should still fail since our own
+        // device is still unverified.
+        let keys_query = KeyDistributionTestData::me_keys_query_response();
+        machine.mark_request_as_sent(&TransactionId::new(), &keys_query).await.unwrap();
+
+        let request_result = machine
+            .share_room_key(
+                fake_room_id,
+                iter::once(KeyDistributionTestData::dan_id()),
+                encryption_settings.clone(),
+            )
+            .await;
+
+        assert_matches!(
+            request_result,
+            Err(OlmError::SessionRecipientCollectionError(
+                SessionRecipientCollectionError::SendingFromUnverifiedDevice
+            ))
+        );
+
+        // Finally, after we trust our own cross-signing keys, key sharing
+        // should succeed.
+        machine
+            .import_cross_signing_keys(CrossSigningKeyExport {
+                master_key: KeyDistributionTestData::MASTER_KEY_PRIVATE_EXPORT.to_owned().into(),
+                self_signing_key: KeyDistributionTestData::SELF_SIGNING_KEY_PRIVATE_EXPORT
+                    .to_owned()
+                    .into(),
+                user_signing_key: KeyDistributionTestData::USER_SIGNING_KEY_PRIVATE_EXPORT
+                    .to_owned()
+                    .into(),
+            })
+            .await
+            .unwrap();
+
+        let requests = machine
+            .share_room_key(
+                fake_room_id,
+                iter::once(KeyDistributionTestData::dan_id()),
+                encryption_settings.clone(),
+            )
+            .await
+            .unwrap();
+
+        // Dan has two devices, but only one is cross-signed, so there should
+        // only be one key share.
+        assert_eq!(requests.len(), 1);
+    }
+
+    #[async_test]
+    async fn test_share_identity_strategy_report_pinning_violation() {
+        // Test that identity-based key sharing gives an error when a verified
+        // user changes their identity.
+        let machine: OlmMachine = OlmMachine::new(
+            KeyDistributionTestData::me_id(),
+            KeyDistributionTestData::me_device_id(),
+        )
+        .await;
+
+        machine.bootstrap_cross_signing(false).await.unwrap();
+
+        let keys_query = IdentityChangeDataSet::key_query_with_identity_a();
+        machine.mark_request_as_sent(&TransactionId::new(), &keys_query).await.unwrap();
+
+        let keys_query = MaloIdentityChangeDataSet::initial_key_query();
+        machine.mark_request_as_sent(&TransactionId::new(), &keys_query).await.unwrap();
```

Comment:
looks like we're mixing up data from two different test sets. That's okay, if that test data is the most appropriate for this test, but could you add a comment in here to explain what's going on? Which users' keys are we importing here?

Reply:
Added some comments.  I also noticed that we were basically just doing the same thing to both users, so I changed it so that we withdraw verification on one user, and re-verify the second user, so that it tests both ways of resolving the issue.